torch/_dynamo/symbolic_convert.py

```diff
@@ -139,6 +139,7 @@ class SpeculationEntry:
     lineno: int
     instruction_pointer: int
     failed: bool = False
+    is_sound: bool = False
```

Comment:
Wait, why do we need both this and failed?

Reply:
I thought having separate flags would be easier for debugging vs having the same flag being set in two different contexts.

Reply:
I could have a single flag set with properties, but it felt more complicated than having an extra flag tbh.

Reply:
So two bools leads to 4 difference cases.  What are those for 4 cases and why do they need to exist?

Reply:
The flags are used in mutually exclusive contexts, so there's really only two cases. They shouldn't be set at the same time.

Reply:
I don't see why we need to restart analysis in the case of things passing.

Also, I don't think sound is the right word here, sound is a property of a technique.

Reply:
cc @voznesenskym can you elaborate on why we needed always restore? and why we need to throw away the fwd/bwd trace? 

My understanding is that we verify that fwd + bwd have certain properties, if they do, we want to **always** rewind the trace and put **apply** directly in the graph instead. We shouldn't be keeping the trace state from backward. 

I used sound because it's consistent with the autograd function code comments. (I could change those too but I can do that separately)

Reply:
`always_restore = self.value.__name__ == "trampoline_autograd_bwd"` 

We should only be always restoring backwards. Forwards restores if it graph breaks. Backwards must always be restored because we don't want side effects from speculating backwards in the graph.